packages/loopring_v3/contracts/impl/LoopringV3.sol

```diff
@@ -16,68 +16,103 @@
 */
 pragma solidity 0.5.10;
 
-import "../iface/ILoopringV3.sol";
-import "../iface/IExchange.sol";
-
+import "../lib/AddressUtil.sol";
 import "../lib/BurnableERC20.sol";
-import "../lib/Claimable.sol";
 import "../lib/ERC20SafeTransfer.sol";
 import "../lib/MathUint.sol";
 
-import "./ExchangeDeployer.sol";
+import "../iface/IExchangeV3.sol";
+import "../iface/ILoopringV3.sol";
 
+import "./ExchangeV3Deployer.sol";
 
-/// @title An Implementation of ILoopring.
+
+/// @title LoopringV3
+/// @dev This contract does NOT support proxy.
 /// @author Brecht Devos - <brecht@loopring.org>
 /// @author Daniel Wang  - <daniel@loopring.org>
-contract LoopringV3 is ILoopringV3, Claimable
+contract LoopringV3 is ILoopringV3
 {
     using AddressUtil       for address payable;
     using MathUint          for uint;
     using ERC20SafeTransfer for address;
 
     // -- Constructor --
     constructor(
-        address payable _protocolFeeVault,
         address _lrcAddress,
         address _wethAddress,
+        address payable _protocolFeeVault,
         address _blockVerifierAddress,
-        address _downtimeCostCalculator,
-        uint    _exchangeCreationCostLRC,
-        uint    _maxWithdrawalFee,
-        uint    _tokenRegistrationFeeLRCBase,
-        uint    _tokenRegistrationFeeLRCDelta,
-        uint    _minExchangeStakeWithDataAvailability,
-        uint    _minExchangeStakeWithoutDataAvailability,
-        uint    _revertFineLRC,
-        uint    _withdrawalFineLRC
+        address _downtimeCostCalculator
         )
+        Claimable()
         public
     {
-        require(address(0) != _protocolFeeVault, "ZERO_ADDRESS");
         require(address(0) != _lrcAddress, "ZERO_ADDRESS");
         require(address(0) != _wethAddress, "ZERO_ADDRESS");
 
-        protocolFeeVault = _protocolFeeVault;
         lrcAddress = _lrcAddress;
         wethAddress = _wethAddress;
 
         updateSettingsInternal(
+            _protocolFeeVault,
             _blockVerifierAddress,
             _downtimeCostCalculator,
-            _exchangeCreationCostLRC,
-            _maxWithdrawalFee,
-            _tokenRegistrationFeeLRCBase,
-            _tokenRegistrationFeeLRCDelta,
-            _minExchangeStakeWithDataAvailability,
-            _minExchangeStakeWithoutDataAvailability,
-            _revertFineLRC,
-            _withdrawalFineLRC
+            0, 0, 0, 0, 0, 0, 0, 0
+        );
+    }
+
+    // === ILoopring methods ===
+    function createExchange()
+        external
+        nonReentrant
+        returns (address exchange)
+    {
+        exchange = ExchangeV3Deployer.deploy();
+        emit ExchangeCreated(exchange);
+    }
+
+
+    function initializeExchange(
+        address exchangeAddress,
+        uint    exchangeId,
```

Comment:
Instead of asking caller to pass into an exchangeId, why not we maintain it in LoopringV3 contract?

Reply:
Currently, we have only one major version (LoopringV3), but in the future, we may have multiple ones. All of them should be registered in the same ProtocolRegistry. Therefore the ID is global to ProtocolRegistry, not LoopringV3.